File: punkc/src/env.ml
open Utils

module StringMap = Map.Make(String)
module Int = struct type t = int let compare = compare end
module IntMap = Map.Make(Int)

type context =
  { ksize: int; kctx: Ast.kind list; tctx: (int * Ast.ty) IntMap.t }

type env = { var_id_map: int StringMap.t;
             ctx: context;
             is_top: bool;
             mut_set: (int, unit) Hashtbl.t;
             elab_con_map: (int, Tir.con) Hashtbl.t;
             persistent_set: (int, unit) Hashtbl.t }

let empty_ctx () = { ksize = 0; kctx = []; tctx = IntMap.empty }

let empty_env () = { var_id_map = StringMap.empty;
                     ctx = empty_ctx ();
                     is_top = true;
                     mut_set = Hashtbl.create table_size;
                     elab_con_map = Hashtbl.create table_size;
                     persistent_set = Hashtbl.create table_size }

let add_id env id name =
  { env with var_id_map = StringMap.add name id env.var_id_map; }

let find_id_opt env name = StringMap.find_opt name env.var_id_map

let lookup_kind ({ kctx }: context) i =
  Subst.lift_kind (i + 1) (List.nth kctx i)

let lookup_type ({ ksize; tctx }: context) v =
  match IntMap.find_opt v tctx with
  | Some (n, c) -> Subst.lift_con (ksize - n) c
  | None -> raise (TypeError ("lookup_type: " ^ (string_of_int v)))

let extend_kind { ksize; kctx; tctx } k =
  { ksize = ksize + 1; kctx = k::kctx; tctx = tctx }

let extend_type { ksize; kctx; tctx } v c =
  { ksize = ksize; kctx = kctx; tctx = IntMap.add v (ksize, c) tctx}

let ksize ({ ksize }: context) = ksize

let new_func_name () =
  let (id, _) = Var.newvar None in
  "f" ^ string_of_int id

let mangle_name id = "v" ^ string_of_int id

File: punkc_fs/frontend/env.fs
module Env

open Microsoft.FSharp.Collections
open System.Collections.Generic
open Config
open Errors

type context =
    { ksize : int
      kctx : ir.Ir.kind list
      tctx : Dictionary<int, int * ir.Ir.ty>
      struct_map : Dictionary<int, ir.Ir.kind> }

type env =
    { var_id_map : Map<string, int>
      ctx : context
      is_top : bool
      elab_con_map : Dictionary<int, Tir.con>
      struct_def : Dictionary<int, (string * ir.Ir.ty) list> }

let emptyCtx() =
    { ksize = 0
      kctx = []
      tctx = new Dictionary<_, _>()
      struct_map = new Dictionary<_, _>() }

let emptyEnv() =
    { var_id_map = Map.empty
      ctx = emptyCtx()
      is_top = true
      elab_con_map = new Dictionary<_, _>()
      struct_def = new Dictionary<_, _>() }

let addId env id name =
    { env with var_id_map = Map.add name id env.var_id_map }

let tryFindId env name = Map.tryFind name env.var_id_map

let lookupKind (ctx : context) i = List.item i ctx.kctx

let lookupType (ctx : context) v =
    match ctx.tctx.TryGetValue(v) with
    | true, (n, c) -> Subst.liftCon (ctx.ksize - n) c
    | false, _ -> raise (TypeError("TODO lookupType"))

let lookupStruct (ctx : context) i =
    match ctx.struct_map.TryGetValue(i) with
    | false, _ -> raise (TypeError "TODO lookupStruct")
    | true, v -> v

let extendKind (ctx : context) k =
    { struct_map = ctx.struct_map
      ksize = ctx.ksize + 1
      kctx = k :: ctx.kctx
      tctx = ctx.tctx }

let extendType (ctx : context) v c =
    ctx.tctx.Add(v, (ctx.ksize, c))

let ksize (ctx : context) = ctx.ksize
